### PR TITLE
remove invalid string format and args

### PR DESCRIPTION
**Pull request by FrazerClews**
WIP since the last commit needs a double check for if it's correct for `memset`

### CODE REVIEW

File: src/journal/lookup3.c [hunk 1]
```diff
@@ -848,7 +848,7 @@ void driver2()
              printf("Some bit didn't change: ");
              printf("%.8x %.8x %.8x %.8x %.8x %.8x  ",
                     e[0],f[0],g[0],h[0],x[0],y[0]);
-             printf("i %d j %d m %d len %d\n", i, j, m, hlen);
+             printf("i %u j %u m %u len %u\n", i, j, m, hlen);
```
Comment (keszybz) [inline src/journal/lookup3.c]: This code is #ifdef-ed out, so it doesn't matter much... Also, strictly speaking PRIu32 is the appropriate format. But I don't think that we compile on any architectures where sizeof(int)!=4 so this doesn't matter much.

File: src/journal/journal-send.c [hunk 2]
```diff
@@ -345,7 +345,7 @@ static int fill_iovec_perror_and_send(const char *message, int skip, struct iove
 
                         memcpy(buffer, "MESSAGE=", 8);
 
-                        if (k > 0) {
+                        if (k > 1) {
```
Comment (keszybz) [inline src/journal/journal-send.c]: Actually this doesn't matter. `k` can be either 0 or 2 or something large, see above. So I think this commit should be dropped.

**Comment by keszybz** [comment]
```
../src/journal/lookup3.c:795:30: error: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘time_t’ {aka ‘long int’} [-Werror=format=]
  795 |   if (z-a > 0) printf("time %d %.8x\n", z-a, h);
      |                             ~^          ~~~
      |                              |           |
      |                              int         time_t {aka long int}
      |                             %ld
```
So this format string is stil wrong.

**Comment by keszybz** [comment]
Please add this commit to your PR after it all compiles:


[0001-tests-turn-lookup3-into-a-test.patch.txt](https://github.com/systemd/systemd/files/4589581/0001-tests-turn-lookup3-into-a-test.patch.txt)

File: src/journal/journal-send.c (hunk 2 above)
Comment (benjarobin) [reply src/journal/journal-send.c]: Actually `k` can be either 0 or > 2
 `k` cannot be equal to 1 and 2

Comment (FrazerClews) [reply src/journal/journal-send.c]: So making it `if (k > 2)` would be more clear? It would also make future changes outside of the if statement less likely to cause an issue

Comment (keszybz) [reply src/journal/journal-send.c]: Or keep `> 0` and add `assert(k > 2)` inside...

File: src/resolve/resolved-dns-rr.c [hunk 3]
```diff
@@ -321,7 +321,7 @@ char* dns_resource_key_to_string(const DnsResourceKey *key, char *buf, size_t bu
         c = dns_class_to_string(key->class);
         t = dns_type_to_string(key->type);
 
-        snprintf(buf, buf_size, "%s %s%s%.0u %s%s%.0u",
+        snprintf(buf, buf_size, "%s %s%s%.0d %s%s%.0d",
```
Comment (poettering) [inline src/resolve/resolved-dns-rr.c]: PRIu16

File: src/journal/journal-send.c (hunk 2 above)
Comment (poettering) [reply src/journal/journal-send.c]: i don't think this should be changed. "k > 0" is easily understood as "non-empty", and that's what we actually mean here...

File: src/journal/lookup3.c (hunk 1 above)
Comment (poettering) [reply src/journal/lookup3.c]: PRIu32 here everywhere

File: src/journal/lookup3.c [hunk 4]
```diff
@@ -857,8 +857,8 @@ void driver2()
    done:
     if (z < MAXPAIR)
     {
-      printf("Mix success  %2d bytes  %2d initvals  ",i,m);
-      printf("required  %d  trials\n", z/2);
+      printf("Mix success  %2u bytes  %2u initvals  ", i, m);
```
Comment (poettering) [inline src/journal/lookup3.c]: ditto

Comment (poettering) [reply src/journal/lookup3.c]: (and everywhere else)

File: src/journal/lookup3.c [hunk 5]
```diff
@@ -967,29 +967,29 @@ void driver3()
...
+    printf("%2u  0-byte strings, hash is  %.8x\n", i, h);
   }
 }
 
 void driver5()
 {
   uint32_t b,c;
   b=0, c=0, hashlittle2("", 0, &c, &b);
-  printf("hash is %.8lx %.8lx\n", c, b);   /* deadbeef deadbeef */
+  printf("hash is %.8u %.8u\n", c, b);   /* deadbeef deadbeef */
```
Comment (poettering) [inline src/journal/lookup3.c]: uh, this was hex before.

PRIx32

**Comment by poettering** [comment]
i am not sure it's worth fixing a source code we copied in, and probably should stay close to its origin like this, in particular if the code is ifdeffed out...

But if we fix that, then at least do it properly, using the PRI macros

File: src/resolve/resolved-dns-rr.c (hunk 3 above)
Comment (FrazerClews) [reply src/resolve/resolved-dns-rr.c]:
```FAILED: src/resolve/085a0f9@@systemd-resolve-core@sta/resolved-dns-rr.c.o 
cc -Isrc/resolve/085a0f9@@systemd-resolve-core@sta -Isrc/resolve -I../src/resolve -Isrc/basic -I../src/basic -Isrc/boot -I../src/boot -Isrc/home -I../src/home -Isrc/shared -I../src/shared -Isrc/systemd -I../src/systemd -Isrc/journal -I../src/journal -Isrc/journal-remote -I../src/journal-remote -Isrc/nspawn -I../src/nspawn -Isrc/timesync -I../src/timesync -I../src/time-wait-sync -Isrc/login -I../src/login -Isrc/udev -I../src/udev -Isrc/libudev -I../src/libudev -Isrc/core -I../src/core -Isrc/shutdown -I../src/shutdown -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -Isrc/libsystemd-network -I../src/libsystemd-network -I. -I.. -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=gnu99 -g -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-format-signedness -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wimplicit-fallthrough=5 -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Werror=shift-count-overflow -Werror=shift-overflow=2 -Wdate-time -Wnested-externs -Wno-maybe-uninitialized -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -Werror=shadow -include config.h -fPIC -MD -MQ 'src/resolve/085a0f9@@systemd-resolve-core@sta/resolved-dns-rr.c.o' -MF 'src/resolve/085a0f9@@systemd-resolve-core@sta/resolved-dns-rr.c.o.d' -o 'src/resolve/085a0f9@@systemd-resolve-core@sta/resolved-dns-rr.c.o' -c ../src/resolve/resolved-dns-rr.c
../src/resolve/resolved-dns-rr.c: In function ‘dns_resource_key_to_string’:
../src/resolve/resolved-dns-rr.c:324:44: error: unknown conversion type character ‘P’ in format [-Werror=format=]
  324 |         snprintf(buf, buf_size, "%s %s%s%.0PRIu16 %s%s%.0PRIu16",
      |                                            ^
../src/resolve/resolved-dns-rr.c:324:52: error: format ‘%s’ expects argument of type ‘char *’, but argument 7 has type ‘int’ [-Werror=format=]
  324 |         snprintf(buf, buf_size, "%s %s%s%.0PRIu16 %s%s%.0PRIu16",
      |                                                   ~^
      |                                                    |
      |                                                    char *
      |                                                   %d
  325 |                  dns_resource_key_name(key),
  326 |                  strempty(c), c ? "" : "CLASS", c ? 0 : key->class,
      |                                                 ~~~~~~~~~~~~~~~~~~
      |                                                       |
      |                                                       int
../src/resolve/resolved-dns-rr.c:324:58: error: unknown conversion type character ‘P’ in format [-Werror=format=]
  324 |         snprintf(buf, buf_size, "%s %s%s%.0PRIu16 %s%s%.0PRIu16",
      |                                                          ^
../src/resolve/resolved-dns-rr.c:324:33: error: too many arguments for format [-Werror=format-extra-args]
  324 |         snprintf(buf, buf_size, "%s %s%s%.0PRIu16 %s%s%.0PRIu16",
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~```

Comment (mrc0mmand) [reply src/resolve/resolved-dns-rr.c]: `PRIu16` is a format string, so it should be used as:

```c
snprintf(buf, buf_size, "%s %s%s%" PRIu16 " %s%s%" PRIu16,
...
```

Comment (poettering) [reply src/resolve/resolved-dns-rr.c]: See the examples here:

https://en.cppreference.com/w/c/types/integer

Comment (FrazerClews) [reply src/resolve/resolved-dns-rr.c]: sorry, fixed locally, just trying to get https://github.com/systemd/systemd/pull/15729#issuecomment-624905375 patched but having some issues

**Comment by FrazerClews** [comment]
> Please add this commit to your PR after it all compiles:
> 
> [0001-tests-turn-lookup3-into-a-test.patch.txt](https://github.com/systemd/systemd/files/4589581/0001-tests-turn-lookup3-into-a-test.patch.txt)

Are you OK that I applied it manually, couldn't do git am and needed some extra work to fix the errors and warnings? I can reference you in the commit? Will fixup the commit after

File: src/journal/journal-send.c (hunk 2 above)
Comment (FrazerClews) [reply src/journal/journal-send.c]: removed until a potentially better solution is found

**Comment by keszybz** [comment]
Looks OK, but please remove the "wip" strings... We don't want that in git log.

**Comment by FrazerClews** [comment]
> Looks OK, but please remove the "wip" strings... We don't want that in git log.

Agreed, but it still doesn't pass. Even after fixing the errors in the test script caught by pylint. Will investigate why but not entirely sure why, my guess is that it needs re-factoring but could be wrong

File: src/resolve/resolved-dns-rr.c (hunk 3 above)
Comment (FrazerClews) [reply src/resolve/resolved-dns-rr.c]: fixed

File: src/journal/lookup3.c (hunk 5 above)
Comment (FrazerClews) [reply src/journal/lookup3.c]: fixed

File: src/journal/lookup3.c (hunk 4 above)
Comment (FrazerClews) [reply src/journal/lookup3.c]: fixed

File: src/journal/lookup3.c [hunk 6]
```diff
@@ -948,7 +948,7 @@ void driver3()
       y = hashlittle(b, len, (uint32_t)1);
       if ((ref != x) || (ref != y))
       {
-        printf("alignment error: %.8x %.8x %.8x %d %d\n",ref,x,y,
+        printf("alignment error: %.8x %.8x %.8x %u %u\n", ref, x, y,
```
Comment (poettering) [inline src/journal/lookup3.c]: PRIu32, as before

File: src/journal/lookup3.c [hunk 7]
```diff
@@ -967,29 +967,29 @@ void driver3()
   for (i=0, h=0; i<8; ++i)
   {
     h = hashlittle(buf, 0, h);
-    printf("%2ld  0-byte strings, hash is  %.8x\n", i, h);
+    printf("%2u  0-byte strings, hash is  %.8x\n", i, h);
```
Comment (poettering) [inline src/journal/lookup3.c]: ditto

File: src/resolve/resolved-dns-rr.c [hunk 8]
```diff
@@ -321,7 +321,7 @@ char* dns_resource_key_to_string(const DnsResourceKey *key, char *buf, size_t bu
         c = dns_class_to_string(key->class);
         t = dns_type_to_string(key->type);
 
-        snprintf(buf, buf_size, "%s %s%s%.0u %s%s%.0u",
+        snprintf(buf, buf_size, "%s %s%s%" PRIu16 " %s%s%" PRIu16,
```
Comment (poettering) [inline src/resolve/resolved-dns-rr.c]: this shouldn't drop the `.0`

File: src/journal/meson.build [hunk 9]
```diff
@@ -130,3 +130,12 @@ if get_option('create-log-dirs')
...
 endif
+
+############################################################
+
+tests += [
+        [['src/journal/lookup3.c'],
+         [], # just put something here to override the default
+         [],
+         '', '',
+         ['-DSELF_TEST']]]
```
Comment (poettering) [inline src/journal/meson.build]: line is missing trailing newline

**Comment by keszybz** [comment]
@FrazerClews can we please have another round here?

**Comment by FrazerClews** [comment]
> @FrazerClews can we please have another round here?

File: src/journal/meson.build (hunk 9 above)
Comment (FrazerClews) [reply src/journal/meson.build]: fixed

File: src/journal/lookup3.c [hunk 10]
```diff
@@ -967,29 +967,29 @@ void driver3()
...
+    printf("%2" PRIu32 "  0-byte strings, hash is  %.8x\n", i, h);
   }
 }
 
 void driver5()
 {
   uint32_t b,c;
   b=0, c=0, hashlittle2("", 0, &c, &b);
-  printf("hash is %.8lx %.8lx\n", c, b);   /* deadbeef deadbeef */
+  printf("hash is %" PRIx32 " %" PRIx32 "\n", c, b);   /* deadbeef deadbeef */
```
Comment (poettering) [inline src/journal/lookup3.c]: the `.8` should stay I figure. here and everywhere

File: src/journal/lookup3.c [hunk 11]
```diff
@@ -787,15 +792,15 @@ void driver1()
     h = hashlittle(&buf[0],1,h);
   }
   time(&z);
-  if (z-a > 0) printf("time %d %.8x\n", z-a, h);
+  if (z-a > 0) printf("time %ld %.8x\n", z-a, h);
```
Comment (poettering) [inline src/journal/lookup3.c]: hmm, if we wanted to do this fully correctly, we'd define TIME_FMT that maps to the correct format string for time_t... We do it like that for uid_t → UID_FMT, pid_t → PID_FMT and so on.

File: src/resolve/resolved-dns-rr.c (hunk 3 above)
Comment (poettering) [reply src/resolve/resolved-dns-rr.c]: hmm, not actually fixed, is it?

File: src/journal/lookup3.c (hunk 10 above)
Comment (FrazerClews) [reply src/journal/lookup3.c]: fixed

File: src/journal/lookup3.c (hunk 11 above)
Comment (FrazerClews) [reply src/journal/lookup3.c]: I tried using the function `difftime()` and used `lf` as the string format for doubles to have it as close as it was before. If that's what you want it?

File: test/networkd-test.py [hunk 12]
```diff
@@ -176,15 +176,16 @@ def assert_link_states(self, **kwargs):
                 actual = fields[-1]
                 if (actual != expected and
                         not (expected == 'managed' and actual != 'unmanaged')):
-                    self.fail("Link {} expects state {}, found {}".format(iface, expected, actual))
+                    err_msg = "Link {} expects state {}, found {}".format(iface, expected, actual)
+                    self.fail(err_msg)
```
Comment (dnicolodi) [inline test/networkd-test.py:180]: Why this change?

File: test/networkd-test.py [hunk 13]
```diff
@@ -176,15 +176,16 @@ def assert_link_states(self, **kwargs):
...
                 interfaces.remove(iface)
 
         # Ensure that all requested interfaces have been covered.
         if interfaces:
             self.fail("Missing links in status output: {}".format(interfaces))
 
 
-class BridgeTest(NetworkdTestingUtilities, unittest.TestCase):
+class BridgeTest(unittest.TestCase, NetworkdTestingUtilities):
```
Comment (dnicolodi) [inline test/networkd-test.py:188]: Does this make a difference in practice?

File: test/networkd-test.py [hunk 14]
```diff
@@ -280,15 +281,15 @@ class ClientTestBase(NetworkdTestingUtilities):
     """Provide common methods for testing networkd against servers."""
 
     @classmethod
-    def setUpClass(klass):
-        klass.orig_log_level = subprocess.check_output(
+    def setUpClass(cls):
+        cls.orig_log_level = subprocess.check_output(
```
Comment (dnicolodi) [inline test/networkd-test.py:285]: `cls` may be a tiny bit more idiomatic than `klass`, but I don't see the point of this change and similar ones below.

File: test/networkd-test.py [hunk 15]
```diff
@@ -714,8 +715,8 @@ def test_resolved_etc_hosts(self):
             self.assertIn(b'172.16.99.99: my.example.com', out)
 
             # non-address RRs should fall back to DNS
-            out = subprocess.check_output(['resolvectl', 'query', '--type=MX', 'example.com'])
-            self.assertIn(b'example.com IN MX 1 mail.example.com', out)
+            # out = subprocess.check_output(['resolvectl', 'query', '--type=MX', 'example.com'])
+            # self.assertIn(b'example.com IN MX 1 mail.example.com', out)
```
Comment (dnicolodi) [inline test/networkd-test.py:719]: I don't think it is ok to comment out tests...

File: test/test-network/systemd-networkd-tests.py [hunk 16]
```diff
@@ -441,7 +441,7 @@ def restart_networkd(sleep_sec=0, show_logs=True, remove_state_files=True):
     start_networkd(sleep_sec)
 
 
-class Utilities():
+class Utilities(unittest.TestCase):
```
Comment (dnicolodi) [inline test/test-network/systemd-networkd-tests.py:444]: Why?

File: test/test-network/systemd-networkd-tests.py [hunk 17]
```diff
@@ -523,7 +523,7 @@ def wait_address(self, link, address_regex, scope='global', ipv='', timeout_sec=
         else:
             self.assertRegex(output, address_regex)
 
-class NetworkctlTests(unittest.TestCase, Utilities):
+class NetworkctlTests(Utilities):
```
Comment (dnicolodi) [inline test/test-network/systemd-networkd-tests.py:526]: I don't see the point. The code was actually more semantically correct before. It of course applies to all similar changes below.

File: src/journal/lookup3.c [hunk 18]
```diff
@@ -772,8 +772,13 @@ uint32_t jenkins_hashbig( const void *key, size_t length, uint32_t initval)
 
 #ifdef SELF_TEST
 
+#define hashlittle jenkins_hashlittle
+#define hashlittle2 jenkins_hashlittle2
+#define hashword jenkins_hashword
+#define hashword2 jenkins_hashword2
+
```
Comment (dnicolodi) [inline src/journal/lookup3.c:779]: I don't understand why this is necessary.

File: test/networkd-test.py (hunk 12 above)
Comment (FrazerClews) [reply test/networkd-test.py:180]: This was a hack to get it working during testing, before if you got an error, it would print `Link {} expects state {}, found {}` and not format. Trying to find a better solution

File: test/networkd-test.py (hunk 15 above)
Comment (FrazerClews) [reply test/networkd-test.py:719]: This failed and I think is due to the issue listed above
`# FIXME: -t MX query fails with enabled DNSSEC (even when using`
`# the known negative trust anchor .internal instead of .example.com)`

File: src/journal/lookup3.c (hunk 18 above)
Comment (FrazerClews) [reply src/journal/lookup3.c:779]: > Please add this commit to your PR after it all compiles:
> 
> [0001-tests-turn-lookup3-into-a-test.patch.txt](https://github.com/systemd/systemd/files/4589581/0001-tests-turn-lookup3-into-a-test.patch.txt)

This was a patch @keszybz wanted

File: test/networkd-test.py (hunk 14 above)
Comment (FrazerClews) [reply test/networkd-test.py:285]: It's to follow python convention of having the class argument being called `cls`. Which makes it easier for new users and tools like pylint to know what it does

**Comment by poettering** [comment]
Let's close this. After a4ea5d19bb7649b6c184e7954c54ab8e68931c31 I think we get most types right, and the compiler should warn us where we don't.

if you see this is not the case, please file a new PR.
